main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1496,13 +1516,99 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
       case _ => ResolutionError.IllegalTypeApplication(tpe1, tpe2, loc).toFailure
     }
   }
 
+  /**
+    * Create a well-formed type applying `tpe` to `args`.
+    */
+  private def mkApply(tpe: Type, args: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    fold(args, tpe)(mkApply(_, _, loc))
+  }
+
+  /**
+    * Create a well-formed `And` type.
+    */
+  private def mkAnd(tpe1: Type, tpe2: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.And, List(tpe1, tpe2), loc)
+  }
+
+  /**
+    * Create a well-formed `Or` type.
+    */
+  private def mkOr(tpe1: Type, tpe2: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Or, List(tpe1, tpe2), loc)
+  }
+
+  /**
+    * Create a well-formed `Not` type.
+    */
+  private def mkNot(tpe: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Not, tpe, loc)
+  }
+
+  /**
+    * Create a well-formed `Nullable` type.
+    */
+  private def mkNullable(tpe: Type, nullity: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Cst(TypeConstructor.Nullable), List(tpe, nullity), loc)
+  }
+
+  /**
+    * Creates a well-formed `Tuple` type.
+    */
+  private def mkTuple(tpes: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Cst(TypeConstructor.Tuple(tpes.length)), tpes, loc)
+  }
+
+  /**
+    * Creates a well-formed `RecordExtend` type.
+    */
+  private def mkRecordExtend(label: String, tpe: Type, rest: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Cst(TypeConstructor.RecordExtend(label)), List(tpe, rest), loc)
+  }
+
+  /**
+    * Creates a well-formed `SchemaExtend` type.
+    */
+  private def mkSchemaExtend(name: String, tpe: Type, rest: Type, loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkApply(Type.Cst(TypeConstructor.SchemaExtend(name)), List(tpe, rest), loc)
+  }
+
+  /**
+    * Creates a well-formed `Lattice` or `Relation` type.
+    */
+  private def mkPredicate(den: Ast.Denotation, ts0: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    val tycon = den match {
+      case Denotation.Relational => Type.Relation
+      case Denotation.Latticenal => Type.Lattice
+    }
+    val tsVal = ts0 match {
+      case Nil => Type.Unit.toSuccess
+      case x :: Nil => x.toSuccess
+      case xs => mkTuple(xs, loc)
+    }
+
+    flatMapN(tsVal)(mkApply(tycon, _, loc))
+  }
+
+  /**
+    * Creates a well-formed `Relation` type.
+    */
+  private def mkRelation(ts0: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkPredicate(Ast.Denotation.Relational, ts0, loc)
+  }
+
+  /**
+    * Creates a well-formed `Lattice` type.
+    */
+  private def mkLattice(ts0: List[Type], loc: SourceLocation): Validation[Type, ResolutionError] = {
+    mkPredicate(Ast.Denotation.Latticenal, ts0, loc)
+  }
```

Comment:
All of this looks reasonable. Presumably, in the future, if/when we are able to implement kind inference this would be unified in some central place during type inference. However, for now it is useful to have it.